observability/src/test/java/io/grpc/observability/LoggingChannelProviderTest.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.AdditionalAnswers.delegatesTo;
+import static org.mockito.ArgumentMatchers.same;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Grpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.ManagedChannelProvider;
+import io.grpc.MethodDescriptor;
+import io.grpc.TlsChannelCredentials;
+import io.grpc.observability.interceptors.LoggingChannelInterceptor;
+import io.grpc.testing.TestMethodDescriptors;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentMatchers;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public class LoggingChannelProviderTest {
+  @Rule
+  public final MockitoRule mocks = MockitoJUnit.rule();
+
+  private final MethodDescriptor<Void, Void> method = TestMethodDescriptors.voidMethod();
+
+  @Test
+  public void initTwiceCausesException() {
+    LoggingChannelProvider.init(new LoggingChannelInterceptor.FactoryImpl());
+    assertThat(LoggingChannelProvider.instance).isNotNull();
+    ManagedChannelProvider prevProvider = LoggingChannelProvider.instance.prevProvider;
+    assertThat(ManagedChannelProvider.provider()).isSameInstanceAs(LoggingChannelProvider.instance);
+    try {
+      LoggingChannelProvider.init(new LoggingChannelInterceptor.FactoryImpl());
+      fail("should have failed for calling init() again");
+    } catch (IllegalStateException e) {
+      assertThat(e).hasMessageThat().contains("LoggingChannelProvider already initialized!");
+    }
+    LoggingChannelProvider.finish();
+    assertThat(ManagedChannelProvider.provider()).isSameInstanceAs(prevProvider);
```

Comment:
This seems too involved with implementation details. We can just assert that `ManagedChannelProvider.provider()` is no longer `LoggingChannelProvider.instance`. See [don't overuse `@VisibleForTesting`](https://g3doc.corp.google.com/java/com/google/android/gmscore/integ/policy_sets/people/g3doc/devguide/best_practices/index.md#visiblefortesting)

Reply:
```java
    ManagedChannelProvider originalProvider = ManagedChannelProvider.provider();
    LoggingChannelProvider.init(new LoggingChannelInterceptor.FactoryImpl());
    assertThat(LoggingChannelProvider.instance).isNotNull();
    assertThat(ManagedChannelProvider.provider()).isNotSameInstanceAs(originalProvider);
    try {
      LoggingChannelProvider.init(new LoggingChannelInterceptor.FactoryImpl());
      fail("should have failed for calling init() again");
    } catch (IllegalStateException e) {
      assertThat(e).hasMessageThat().contains("LoggingChannelProvider already initialized!");
    }
    LoggingChannelProvider.finish();
    assertThat(ManagedChannelProvider.provider()).isSameInstanceAs(originalProvider);
```

Reply:
Regarding this:

```
assertThat(ManagedChannelProvider.provider()).isNotSameInstanceAs(originalProvider);
```

instead of 

```
assertThat(ManagedChannelProvider.provider()).isSameInstanceAs(LoggingChannelProvider.instance);
```

if `ManagedChannelProvider.provider()` is returning a different provider after `LoggingChannelProvider` is registered then it has to be a `LoggingChannelProvider`. So why do you think it's a bad idea to check for the top provider to be the `LoggingChannelProvider.instance`?

Reply:
Just nitpick. We can assert that  `instanceof LoggingChannelProvider`. While asserting it is the same instance as `LoggingChannelProvider.instance` is not bad, just not testing with the API interface but it's field.

Reply:
Yes, I thought of that. I will change it to instanceof

Reply:
Done

Reply:
done